File: javascript/binary-search-tree/binary-search-tree.js
const Bst = function(data) {
  this.data = data;
  this.left;
  this.right;
};

Bst.prototype.insert = function(value) {
  let insertTo = 'right';
  if (value <= this.data) {
    insertTo = 'left';
  }
  if (!this[insertTo]) {
    this[insertTo] = new Bst(value);
  } else {
    this[insertTo].insert(value);
  }
  return this;
};

Bst.prototype.each = function(f) {
  if (this.left) { this.left.each(f); }
  f(this.data);
  if (this.right) { this.right.each(f); }
};

module.exports = Bst;

File: ruby/binary-search-tree/binary_search_tree.rb
class Bst
  attr_reader :data, :left, :right
  def initialize(value)
    @data  = value
    @left  = nil
    @right = nil
  end

  def insert(value)
    if value <= @data
      @left.insert(value) if @left != nil
      @left = Bst.new(value) if @left == nil
    else
      @right.insert(value) if @right != nil
      @right = Bst.new(value) if @right == nil
    end
  end

  def each(&block)
    return enum_for(:each) unless block_given?
    left.each(&block) if @left
    yield(@data)
    right.each(&block) if @right
  end
end

module BookKeeping
  Bst::VERSION = 1
end
